activerecord/lib/active_record/associations/preloader/batch.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module Associations
+    class Preloader
+      class Batch #:nodoc:
+        def initialize(preloaders)
+          @preloaders = preloaders.reject(&:empty?)
+        end
+
+        def call
+          return if @preloaders.empty?
+
+          loaders = @preloaders.flat_map(&:loaders)
+          group_and_load_similar(loaders)
+          loaders.each(&:run)
+
+          child_preloaders = @preloaders.flat_map(&:child_preloaders)
+          Batch.new(child_preloaders).call
+        end
+
+        private
+          attr_reader :loaders
+
+          def group_and_load_similar(loaders)
+            loaders.grep_v(ThroughAssociation).group_by(&:grouping_key).each do |(_, _, association_key_name), similar_loaders|
```

Comment:
this `grep_v` broke some tests in our app and we are currently trying to understand why N+1 queries are happening since the last change. @byroot has more info, but just wanted to give a heads up that the behavior of the preloader changed if you know why would do that.